### PR TITLE
document serializability in the Templates reference doc

### LINKED ISSUE #15331: define serializable type in user docs
@carrielaben-da suggests either an explanation or a link to an explanation of the 'several other restrictions' of the definition of "serializable types" if they are likely to be things that the reader will actually encounter.

_Originally posted by @carrielaben-da in https://github.com/digital-asset/daml/pull/15328#discussion_r1003679102_

### PR DESCRIPTION
**Pull request by S11001001**
Fixes #15331.

### CODE REVIEW

**Review by ray-roestenburg-da — APPROVED** [review]

File: docs/source/daml/reference/templates.rst [hunk 1]
```diff
@@ -137,6 +137,31 @@ Choices
   Starting with ``choice`` lets you pass in a ``Party`` to use as a controller. But you must make sure to add that party as an ``observer``.
 - See :doc:`choices` for full reference information.
 
+Serializable Types
+******************
+
+Every parameter to a template, choice argument and choice result must have a *serializable type*.
```
Comment (katmurp) [inline docs/source/daml/reference/templates.rst:143]:
```suggestion
Every parameter to a template, choice argument, and choice result must have a *serializable type*.
```

File: docs/source/daml/reference/templates.rst [hunk 2]
```diff
@@ -137,6 +137,31 @@ Choices
...
+This does not merely mean "convertible to bytes"; it has a specific meaning in Daml.
+The serializability rule serves three purposes:
+
+1. offer a stable means to store ledger values permanently,
```
Comment (katmurp) [inline docs/source/daml/reference/templates.rst:147]:
```suggestion
1. Offer a stable means to store ledger values permanently.
```

File: docs/source/daml/reference/templates.rst [hunk 3]
```diff
@@ -137,6 +137,31 @@ Choices
...
+2. provide a sensible encoding of them over :doc:`/app-dev/ledger-api`, and
+3. provide sensible *types* that directly match their Daml counterparts in languages like Java for language codegen.
```
Comment (katmurp) [inline docs/source/daml/reference/templates.rst:149]:
```suggestion
3. Provide sensible *types* that directly match their Daml counterparts in languages like Java for language codegen.
```

File: docs/source/daml/reference/templates.rst [hunk 4]
```diff
@@ -137,6 +137,31 @@ Choices
...
+
+For example, certain kinds of type parameters Daml offers are perfectly compatible with (1) and (2), but have no proper counterpart in (3), so they are disallowed.
+Similarly, function types have a perfectly sensible Java counterpart, satisfying (3), but no reliable way to store or share them via the API, thus failing (1) and (2).
```
Comment (katmurp) [inline docs/source/daml/reference/templates.rst:152]:
```suggestion
Similarly, function types have sensible Java counterparts, satisfying (3), but no reliable way to store or share them via the API, thus failing (1) and (2).
```

File: docs/source/daml/reference/templates.rst [hunk 5]
```diff
@@ -137,6 +137,31 @@ Choices
...
+The following types are *not serializable*, and thus may not be used in templates.
+
+1. function types
+2. record types where any field is not serializable
+3. variant types where any value is not serializable
+4. variant and enum types with no constructors
+5. references to a parameterized data type where any type argument is not serializable.
+   This applies whether or not the data type definition uses the type parameter
+6. defined data types with type parameters of kind ``Nat``, or any kind other than ``*``.
+   This means higher-kinded types and types that take a parameter just to pass to ``Numeric`` are not serializable
```
Comment (katmurp) [inline docs/source/daml/reference/templates.rst:163]:
```suggestion

- Function types.
- Record types where no field is serializable.
- Variant types where no value is serializable.
- Variant and enum types with no constructors.
- References to a parameterized data type where no type argument is serializable. This applies whether or not the data type definition uses the type parameter.
- Defined data types with type parameters of kind ``Nat``, or any kind other than ``*``. This means higher-kinded types, and types that take a parameter just to pass to ``Numeric``, are not serializable.
```

Comment (S11001001) [reply docs/source/daml/reference/templates.rst:163]: > Record types where no field is serializable.

Would not be accurate; a record with one serializable field and one non-serializable field is non-serializable, despite not fitting the quoted change.

Comment (katmurp) [reply docs/source/daml/reference/templates.rst:163]: Yeah, makes sense. I wonder if there's a clearer way of saying it, perhaps "where there is any non-serializable field"?

Comment (katmurp) [reply docs/source/daml/reference/templates.rst:163]: I would definitely capitalize, add full stops, and put bullets instead of numbers.

Comment (S11001001) [reply docs/source/daml/reference/templates.rst]: 42754fb4e5d013d7c60fe56af22d1b56f7a48946

**Pull request merged**